keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,903 @@
+[topolvm-scheduler](https://github.com/cybozu-go/topolvm/blob/master/docs/design.md#how-the-scheduler-extension-works)
+currently does that with a driver-specific way of storing capacity
+information.
+
+#### Operators for applications
+
+Application operators for modern scale out storage services
+(e.g. MongoDB, ElasticSearch, Kafka, MySQL, PostgreSQL, Minio, etc.)
+may want to control creation of volumes carefully in order to optimize
+availability, durability, performance and cost. For more information
+about this, see the [StoragePool API for Advanced Storage Placement
+KEP](https://github.com/kubernetes/enhancements/pull/1347).
+
+Storage pools as introduced in this KEP enable the creation of such
+operators by providing them the necessary information about storage in
+the cluster.
+
+### Caching remaining capacity via the API server
+
+CSI defines the `GetCapacity` RPC call for the controller service to
+report information about the current capacity but it's not accessible
+in the Kubernetes scheduler. A new type `CSIStoragePool` type gets
+added to store the result of the `GetCapacity` calls for one storage
+pool in the API server and thus make it available to the scheduler. In
+the future it might get extended to also store other information like
+health of the pool.
+
+### Identifying storage pools
+
+A method for identifying storage pools accessible via a CSI driver
+based on the existing topology information gets outlined below. The
+`external-provisioner` will be extended to handle the management of
+`CSIStoragePool` objects.
+
+For local storage in the node, there will be one such `CSIStoragePool`
+object per node and CSI driver on that node. For network attached
+storage, there will be one `CSIStoragePool` per CSI
+[Topology](https://github.com/container-storage-interface/spec/blob/4731db0e0bc53238b93850f43ab05d9355df0fd9/lib/go/csi/csi.pb.go#L1662-L1691)
+and driver.
+
+### Size of ephemeral inline volumes
+
+Currently Kubernetes has no information about the size of an ephemeral
+inline volume. A new `CSIVolumeSource.fsSize` field needs to be added
+to expose that in a vendor-agnostic way. Details for that are in
+https://github.com/kubernetes/enhancements/pull/1353.
+
+### Pod scheduling
+
+The Kubernetes scheduler monitors `CSIStoragePool` objects and
+excludes nodes with insufficient remaining capacity when it comes to
+making scheduling decisions for a pod which uses ephemeral inline
+volumes or persistent volumes with delayed binding. If the driver does
+not indicate that it supports capacity reporting, then the scheduler
+proceeds just as it does now, so nothing changes for existing CSI
+driver deployments.
+
+### API
+
+All Kubernetes API changes are under the `CSIStorageCapacity` feature
+gate.
+
+#### CSIStoragePool
+
+The prefix “CSI” is used to avoid confusion with other types that have
+a similar name and because some aspects are specific to CSI.
+
+```
+// CSIStoragePool identifies one particular storage pool and
```

Comment:
I understand `CSIStoragePool` is something that provides capacity of per node per driver and per storage class.
Currently, there exists `CSINode` which is per node per driver.
How about adding capacity info per storage class to status of `CSINode`?
Scheduler can look at the capacity related info from status of CSINodes related to a driver, which have info per storage class.

Reply:
> I understand CSIStoragePool is something that provides capacity of per node per driver and per storage class.

Not necessarily. The storage pool might be available from multiple different nodes and then replicating the same information for all nodes would be redundant. The "network attached storage" covers this case:

https://github.com/kubernetes/enhancements/blob/a68f768c4c0e401ad45bc15ac2a29391a7d07e97/keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md#example-network-attached-storage)

Reply:
ok.. understood your point wrt 'network attached storage' and reporting capacity.
In this case of NAS, if I'm not wrong, all the nodes given in topology of SC will have same capacity available. If so, then, for this driver, there is no need to be any capacity aware scheduling.

Reply:
Correct, a NAS which is accessible from the entire cluster wouldn't need it. But consider a cluster that spans different data centers and storage that (for whatever reason) is only available to nodes in the same data center as the storage servers. Then one has to track storage capacity per data center.

Reply:
thanks..